lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1167,15 +1167,30 @@ Value *SPIRVToLLVM::mapValue(SPIRVValue *BV, Value *V) {
   if (Loc != ValueMap.end()) {
     if (Loc->second == V)
       return V;
+
+    // Existing LLVM mapped value could be the local SPIRV Builtin variable
+    // created during forward translation. In this case, we are trying to
+    // replace with mapped value with the global SPIRV Builtin variable
+    // (passed via V here). In this case, we also need a AddrSpaceCast to cast
+    // between the new mapped value and the old.
+    if (Loc->second->hasName() &&
+        Loc->second->getName().starts_with("__local_SPIRV_Builtin")) {
+      auto *NewV = new AddrSpaceCastInst(V, Loc->second->getType(), "",
+                                         cast<Instruction>(Loc->second));
+      ValueMap[BV] = NewV;
+      return NewV;
+    }
     auto *LD = dyn_cast<LoadInst>(Loc->second);
-    auto *Placeholder = dyn_cast<GlobalVariable>(LD->getPointerOperand());
-    assert(LD && Placeholder &&
-           Placeholder->getName().starts_with(KPlaceholderPrefix) &&
-           "A value is translated twice");
-    // Replaces placeholders for PHI nodes
-    LD->replaceAllUsesWith(V);
-    LD->eraseFromParent();
-    Placeholder->eraseFromParent();
+    if (LD) {
```

Comment:
This change is not related to this PR. It just adds a possibly missing check.

Reply:
The original code has an assertion that LD is not null, so this check should not be necessary unless the original assert was incorrect.

Reply:
Yes. But the assert was placed after the LD was dereferenced in the previous line.

Reply:
Oops.  Yes, I see.